sdk/ai/azure-ai-inference/src/main/java/com/azure/ai/inference/ChatCompletionClientTracer.java

```diff
@@ -0,0 +1,606 @@
+
+    private void traceChoiceEvent(String choiceContent, OffsetDateTime timestamp, Context span) {
+        final Map<String, Object> eventAttributes = new HashMap<>(2);
+        eventAttributes.put("gen_ai.system", INFERENCE_GEN_AI_SYSTEM_NAME);
+        eventAttributes.put("gen_ai.event.content", choiceContent);
+        tracer.addEvent("gen_ai.choice", eventAttributes, timestamp, span);
+    }
+
+    private String toJsonString(ChatRequestMessage message) {
+        try (ByteArrayOutputStream stream = new ByteArrayOutputStream();
+            JsonWriter writer = JsonProviders.createWriter(stream)) {
+            message.toJson(writer);
+            writer.flush();
+            return new String(stream.toByteArray(), StandardCharsets.UTF_8);
+        } catch (IOException e) {
+            logger.atWarning().log("'ChatRequestMessage' serialization error", e);
+        }
+        return null;
+    }
+
+    private String toJsonString(ChatChoice choice) {
+        try (ByteArrayOutputStream stream = new ByteArrayOutputStream();
+            JsonWriter writer = JsonProviders.createWriter(stream)) {
+            writer.writeStartObject();
+            writer.writeStartObject("message");
+            if (captureContent) {
+                writer.writeStringField("content", choice.getMessage().getContent());
+            }
+            if (choice.getMessage() != null) {
+                final List<ChatCompletionsToolCall> toolCalls = choice.getMessage().getToolCalls();
+                if (toolCalls != null && !toolCalls.isEmpty()) {
+                    writer.writeStartArray("tool_calls");
+                    for (ChatCompletionsToolCall toolCall : toolCalls) {
+                        if (captureContent) {
+                            toolCall.toJson(writer);
+                        } else {
+                            writer.writeStartObject();
+                            writer.writeStringField("id", toolCall.getId());
+                            writer.writeStringField("type", toolCall.getType());
+                            writer.writeEndObject();
+                        }
+                    }
+                    writer.writeEndArray();
+                }
+            }
+            writer.writeEndObject();
+            final CompletionsFinishReason finishReason = choice.getFinishReason();
+            if (finishReason != null) {
+                writer.writeStringField("finish_reason", finishReason.getValue());
+            }
+            writer.writeIntField("index", choice.getIndex());
+            writer.writeEndObject();
+            writer.flush();
+            return new String(stream.toByteArray(), StandardCharsets.UTF_8);
+        } catch (IOException e) {
+            logger.atWarning().log("'ChatChoice' serialization error", e);
+        }
+        return null;
+    }
+
+    private static String getFinishReasons(List<ChatChoice> choices) {
+        final StringJoiner finishReasons = new StringJoiner(",", "[", "]");
```

Comment:
nice, I did not know about this util!

Reply:
yes :) recently learned about this type, and I have been using `StringBuilder` for similar cases.